engine/schema/src/com/cloud/vm/dao/NicDaoImpl.java

```diff
@@ -222,11 +222,20 @@ public String getIpAddress(long networkId, long instanceId) {
     }
 
     @Override
-    public int countNics(long instanceId) {
-        SearchCriteria<Integer> sc = CountBy.create();
-        sc.setParameters("vmId", instanceId);
-        List<Integer> results = customSearch(sc, null);
-        return results.get(0);
+    public int getFreeDeviceId(long instanceId) {
+        Filter searchFilter = new Filter(NicVO.class, "deviceId", true, null, null);
+        SearchCriteria<Integer> sc = deviceIdSearch.create();
+        sc.setParameters("instance", instanceId);
+        List<Integer> deviceIds = customSearch(sc, searchFilter);
+
+        int freeDeviceId = 0;
+        for (int deviceId : deviceIds) {
```

Comment:
@borisroman 
we get the list of deviceId order by asc at first, eg, 0,1,2,4
then we check the list, for index i, if (deviceIds[i] != i), then i is the first free ip.
in my example, the first free device_id is 3 (because deviceIds[3] = 4, not 3)
then we assign 3 as the device_id of new nic.